Central-Server/API/util/communication/packets.py

```diff
@@ -0,0 +1,500 @@
+        Serializes one packet to a raw data string
+        """
+        packet_dict = {'packet_type': self.packet_type, 'packet_data': self.data, 'use_raw': self.use_raw}
+        string_construct = json.dumps(packet_dict)
+        if(self.use_raw):
+            string_construct +=  "[[raw==>]]" + self.raw_data
+        return string_construct
+
+    def __len__(self) -> int:
+        """Returns the length of the serialized string for this packet"""
+        return len(self.serialize())
+
+    def __str__(self) -> str:
+        """Creates a human-readable version of this packet"""
+        if self.use_raw:
+            if(len(self.raw_data) > 50):
+                return "<packet:" + str(self.packet_type) + "> " + str(self.data) + " [raw: +" + str(len(self.raw_data)) + "c]"
+            else:
+                return "<packet:" + str(self.packet_type) + "> " + str(self.data) + " [raw: " + self.raw_data + "]"
+        else:
+            return "<packet:" + str(self.packet_type) + "> " + str(self.data)
+        
+class DataPacketBuffer:
+    """
+    A utility class for encoding and decoding data streams with packets.
+    """
+    packet_buffer: List[DataPacket] = []
+    input_buffer: List[DataPacket] = []
+
+    def __init__(self, packet_list:List[DataPacket]=[]) -> None:
+        self.packet_buffer = packet_list
+
+    def add(self, packet: DataPacket) -> None:
+        """Adds a packet to the packet buffer"""
+        self.packet_buffer.append(packet)
+
+    def write_buffer_to_channel(self, channel: communication_interface.CommunicationChannel) -> None:
+        """
+        Writes the entire packet buffer to serial
+        """
+        serialized_full: str = self.to_serialized_string()
+        if(len(self.packet_buffer) > 0):
+            serialized_full += "[[pkt_end]]"
+        self.packet_buffer = []
+        if(serialized_full != ""):
+            channel.write(serialized_full)
+
+    def write_packet(packet: DataPacket, channel: communication_interface.CommunicationChannel) -> None:
+        """
+        Writes a single packet to serial
+        WARNING: using `write_packet` twice in a row will cause a deserialization error to be thrown due to a lack of a delimeter.
+        """
+        channel.write(packet.serialize())
+
+    def to_serialized_string(self) -> None:
+        """
+        Serializes the packet buffer
+        """
+        serialized_packets: List[str] = []
+        for packet in self.packet_buffer:
+            serialized_packets.append(packet.serialize())
+        serialized_full: str = "[[pkt_end]]".join(serialized_packets)
+        return serialized_full
+
+    def stream_to_packet_list(stream: str, safe_deserialize:bool=False) -> List[DataPacket]:
+        """
+        Converts serialized packet buffer to a list of packets
+        """
+        serialized_packets = stream.split("[[pkt_end]]")
+        ds_packets: List[DataPacket] = []
+        for ser_pkt in serialized_packets:
+            if(ser_pkt == ""):
```

Comment:
nit: in python u don't need the parens for an if

Reply:
Don't need it, but honestly it looks better